app/src/main/java/hu/vmiklos/plees_tracker/GraphsActivity.kt

```diff
@@ -0,0 +1,340 @@
+        chart.axisLeft.removeAllLimitLines()
+        block() // set your data, etc here
+        chart.invalidate()
+    }
+
+    /** Converts raw day-value points into a [LineDataSet] with default values. */
+    private fun List<Pair<Number, Number>>.toDataSet(label_key: Int, color: Int): LineDataSet {
+        return LineDataSet(
+            map { (day, value) -> Entry(day.toFloat(), value.toFloat()) },
+            getString(label_key)
+        ).apply {
+            lineWidth = if (size > 250) 2f else 3f
+            this.color = color
+            setDrawCircles(false)
+            setDrawCircleHole(false)
+            setDrawValues(false)
+        }
+    }
+
+    /** Generates a cumulative moving average list of the provided day-value points. */
+    private fun List<Pair<Number, Number>>.cumulativeAverage(): List<Pair<Number, Number>> {
+        return mutableListOf(first()).also {
+            subList(1, size).forEach { (day, value) ->
+                val oldAvg = it.last().second.toFloat()
+                val newAvg = ((oldAvg * it.size) + value.toFloat()) / (it.size + 1)
+                it.add(day to newAvg)
+            }
+        }
+    }
+
+    /** Generates a cumulative sum list of the provided day-value points. */
+    private fun List<Pair<Number, Number>>.cumulativeSum(): List<Pair<Number, Number>> {
+        return mutableListOf(first()).also {
+            subList(1, size).forEach { (day, value) ->
+                it.add(day to value.toFloat() + it.last().second.toFloat())
+            }
+        }
+    }
+
+    /** Given a list of sleeps, returns list of day-rating pairs, sorted by day. */
+    private fun List<Sleep>.ratingByDay(): List<Pair<Long, Float>> {
+        return groupSleepsByDay()
+            .map { (day, daySleeps) -> day to daySleeps.avgRating() }
+            .sortedBy { it.first }
+    }
+
+    /** Given a list of sleeps, returns list of day-start of sleep pairs, sorted by day. */
+    private fun List<Sleep>.startOfSleepByDay(): List<Pair<Long, Long>> {
+        return groupSleepsByDay()
+            .map { (day, daySleeps) -> day to daySleeps.earliestSleep().start - day }
+            .sortedBy { it.first }
+    }
+
+    /** Given a list of sleeps, returns list of day-sleep length pairs, sorted by day. */
+    private fun List<Sleep>.lengthPerDay(): List<Pair<Long, Float>> {
+        return groupSleepsByDay()
+            .map { (day, daySleeps) -> day to daySleeps.sumLength() }
+            .sortedBy { it.first }
+    }
+
+    /** Given a list of sleeps, returns the average rating. */
+    private fun List<Sleep>.avgRating(): Float {
+        return sumOf { it.rating }.toFloat() / size
+    }
+
+    /** Given a list of sleeps, returns the start time of the earliest sleep. */
+    private fun List<Sleep>.earliestSleep(): Sleep {
+        // run on output of groupBy (groupSleepsByDay) so by definition not empty
+        return minByOrNull { it.start }!!
```

Comment:
This looks dirty: if you call "or null", then just return that type, so the caller can handle the null return value. I imagine in practice this crashes now if you have no sleeps and you try to call this function.

Reply:
Yeah, it doesn't look the greatest - I chose to use `minByOrNull` because `minBy` is deprecated (they do the same thing). I guess I'll just return an empty `Sleep`, which will probably still blow up the graph in some weird way (but I suppose at least it'll render).